packages/@lwc/template-compiler/src/codegen/helpers.ts

```diff
@@ -45,14 +45,15 @@ export function containsDynamicChildren(children: IRNode[]): boolean {
     return children.some((child) => isElement(child) && isDynamic(child));
 }
 
-export function shouldFlatten(children: IRNode[]): boolean {
+export function shouldFlatten(children: IRNode[], state: State): boolean {
     return children.some(
         (child) =>
             isElement(child) &&
             (isDynamic(child) ||
                 !!child.forEach ||
                 !!child.forOf ||
-                (isTemplate(child) && shouldFlatten(child.children)))
+                (state.renderMode === 'light' && child.tag === 'slot') ||
```

Comment:
I still don't fully understand this logic. can someone explains?

Reply:
We're flattening children if they contain a slot (in Light DOM mode). Basically implementing the approach you outlined here: https://github.com/salesforce/lwc/pull/2317#discussion_r636475871

Reply:
Ok, got it!